### PR TITLE
Adding urllib3 integration

### PR DESCRIPTION
**Pull request by justiniso**
Took a stab at an integration for `urllib3`, which is disabled by default. Before I submit tests, I want to get some feedback. 

A few things:

1. `requests` library uses `urllib3`. Longer-term it might be valuable to rework requests integration to use `urllib3`'s integration. Thoughts?
2. The common use with urllib3 is to use a PoolManager, which abstracts the underlying connection pools to a host. So neither `PoolManager` nor `HTTPConnectionPool` are good candidates for storing config info like service name (unlike requests' `Session`). Right now, the only way to change service names is by using the `split_by_host` config. I'm open to other suggestions for configuring service names.
@P403n1x87 use below

# Integration
This PR adds support for [`urllib3`](http://pypi.datadoghq.com/trace-dev/docs/integrations.html#urllib3).

## Links

- Integration docs: http://pypi.datadoghq.com/trace-dev/docs/integrations.html#urllib3
- Corp docs PR: https://github.com/DataDog/documentation/compare/gabriele.tornetta/urllib3-integration

## Checklist

- [x] Usage and configuration documentation added in `__init__.py`, `docs/index.rst` and `docs/integrations.rst`.
- [x] [Corp docs](https://github.com/Datadog/documentation) PR to add new integration to documentation.
- [x] Span metadata
  - [x] Service (use [`int_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L55) or [`ext_service`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L87)).
  - [x] Span type should be one of [these](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/ext/__init__.py#L7).
  - [x] Resource
  - [x] Measured tag
  - [x] Sample analytics
- [x] Global configuration
  - [x] `ddtrace.config` entry is specified.
  - [x] Environment variables are provided for config options.
- [x] Instance configuration
  - [x] Pin overriding.
  - [x] Service name override (if applicable).
- [ ] ~~Async~~
  - [ ] ~~Span parenting behaves as expected.~~
  - [ ] ~~Context propagation across async contexts.~~
- [x] HTTP
  - [x] Distributed tracing propagation is implemented.
  - [x] Use [`trace_utils.set_http_meta`](https://github.com/DataDog/dd-trace-py/blob/90d1d5981c72ea312c21ac04e5be47521d0f0f2e/ddtrace/contrib/trace_utils.py#L143-L152) to set http tags
- [x] Tests
  - [x] Tests are provided for all of the above.
  - [x] Tests are added to CI (`.circleci/config.yml`).
  - [x] Functionality is maintained from original library.
  - [x] Patch test cases are added (see `test_django_patch.py` for an example).
  - [x] All Python versions that the library supports are tested.
  - [x] All significant library versions (including the latest) are tested. This typically includes every minor release going back a few years.

### CODE REVIEW

File: ddtrace/contrib/urllib3/__init__.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    print(r.status)
+    print(r.data)
+
+The auto-instrumentation of requests can be further tuned by configuring
+urllib3 through the Configuration API:
+
+    from ddtrace import config
+
+    # disable distributed tracing
+    config.urllib3['distributed_tracing'] = False
```
Comment (satiani) [inline ddtrace/contrib/urllib3/__init__.py]: Why would anyone want to disable distributed tracing? is this a pattern we're using in other integrations?

File: ddtrace/contrib/urllib3/patch.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+from ...constants import ANALYTICS_SAMPLE_RATE_KEY
+
+
+# Initialize the default config vars
+config._add(
+    "urllib3",
+    {
+        "service_name": get_env("urllib3", "service_name", "urllib3"),
+        "distributed_tracing": asbool(get_env("urllib3", "distributed_tracing", True)),
+        "analytics_enabled": asbool(get_env("urllib3", "analytics_enabled", False)),
```
Comment (satiani) [inline ddtrace/contrib/urllib3/patch.py]: Is that a pattern we use elsewhere? i.e. you have to enable App Analytics on a library by library basis? That seems strange...

Comment (brettlangdon) [reply ddtrace/contrib/urllib3/patch.py]: Yeah, this is very common across all of our integrations.

When you "enable app analytics" it doesn't enable for all integrations by default (by design).

File: ddtrace/contrib/urllib3/__init__.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/contrib/urllib3/__init__.py]: Yeah, this is common across integrations, but very rarely used.

This is mostly for, if I am only making API calls to third party services and I don't want to send the `X-Datadog-*` headers.

File: ddtrace/contrib/urllib3/patch.py (hunk 2 above)
Comment (justiniso) [reply ddtrace/contrib/urllib3/patch.py]: I wanted to keep the behavior across http integrations (requests/aiohttp) as close as possible. But if something has changed (such as going to GA, happy to change the default)

File: ddtrace/contrib/urllib3/patch.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            parsed_uri.fragment,
+        )
+    )
+
+    tracer = getattr(obj, "datadog_tracer", ddtrace.tracer)
+
+    if not tracer.enabled:
+        return func(*args, **kwargs)
+
+    with tracer.trace("urllib3.urlopen", span_type=http.TYPE) as span:
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py]: The `httplib` and `requests` integrations both use span names that end in `.request`.

@justiniso Do you think it makes sense to be consistent on this naming pattern?

File: tests/contrib/urllib3/test_urllib3.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+
+class TestUrllib3(BaseUrllib3TestCase, BaseTracerTestCase):
+    def test_HTTPConnectionPool_traced(self):
+        """Tests that requests made from the HTTPConnectionPool are traced"""
+        pool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+        # Test a relative URL
+        r = pool.request("GET", "/status/200")
+        assert r.status == 200
+        assert len(self.tracer.writer.pop()) == 1
```
Comment (jd) [inline tests/contrib/urllib3/test_urllib3.py]: Should we check what's inside the trace? At least one field to be sure it's not anything else that created it.

File: tests/contrib/urllib3/test_urllib3.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+            (("POST",), {"url": URL_200, "headers": {"accept": "*"}}),
+            ((), {"method": "POST", "url": URL_200, "headers": {"accept": "*"}}),
+        ]
+
+        for args, kwargs in inputs:
+
+            with self.override_config("urllib3", {}):
+                config.urllib3.http.trace_headers(["accept"])
+                pool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+                out = pool.urlopen(*args, **kwargs)
```
Comment (majorgreys) [inline tests/contrib/urllib3/test_urllib3.py]: In the above tests you are calling `pool.request` but here `urlopen`.

```suggestion
                out = pool.request(*args, **kwargs)
```

File: tests/contrib/urllib3/test_urllib3.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+        spans = self.tracer.writer.pop()
+        assert len(spans) == 0
+
+    def test_double_patch(self):
+        """Ensure that double patch doesn't duplicate instrumentation"""
+        patch()
+        connpool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+        setattr(connpool, "datadog_tracer", self.tracer)
+
+        out = connpool.urlopen("GET", URL_200)
```
Comment (majorgreys) [inline tests/contrib/urllib3/test_urllib3.py]:
```suggestion
        out = connpool.request("GET", URL_200)
```

**Review by majorgreys — DISMISSED** [review]
You'll want to add an entry to `.circleci/config.yml`

File: tests/contrib/urllib3/test_urllib3.py (hunk 5 above)
Comment (justiniso) [reply tests/contrib/urllib3/test_urllib3.py]: Specifically this test is checking the signature is correctly interpreted by the patch function `urlopen` is correct sing that is the function being patched here. Tests should be sure to cover all ways in which the user will use the API, which is why some of the tests at the top are using different classes and functions.

File: tests/contrib/urllib3/test_urllib3.py (hunk 6 above)
Comment (justiniso) [reply tests/contrib/urllib3/test_urllib3.py]: Same case as above, this test is checking a double patch on the instrumented class, which is `HTTPConnectionPool` and the patched function on that class is `urlopen`.

File: ddtrace/contrib/urllib3/patch.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+from ...ext import http
+from ...utils.http import sanitize_url_for_tag
+from ...utils.wrappers import unwrap as _u
+from ...utils.formats import asbool, get_env
+from ...propagation.http import HTTPPropagator
+from ...constants import ANALYTICS_SAMPLE_RATE_KEY
+
+
+# Ports which, if set, will not be used in hostnames/service names
+DROP_PORTS = (80, 443)
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py]: If a user of `urllib3` passes the default port for `HTTPConnectionPool` or `HTTPSConnectionPool`, we would be stripping the port from the url. That seems reasonable though an additional step in our code that might not be needed.

I'm more confused by the situation where these default ports are being set without them being specified by the user. Let's make sure we understand why that's the case. Perhaps its specific to certain versions of the library?

At least with 1.25.7, I only see this happen when using `connection_from_url`:

```
>>> from urllib3 import HTTPConnectionPool
>>> import urllib3
>>> p1 = HTTPConnectionPool('ajax.googleapis.com')
>>> p1.scheme, p1.host, p1.port
('http', 'ajax.googleapis.com', None)
>>> p2 = urllib3.connectionpool.connection_from_url("http://ajax.googleapis.com")
>>> p2.scheme, p2.host, p2.port
('http', 'ajax.googleapis.com', 80)
>>> p3 = urllib3.connectionpool.connection_from_url("https://ajax.googleapis.com")
>>> p3.scheme, p3.host, p3.port
('https', 'ajax.googleapis.com', 443)
>>> from urllib3 import HTTPSConnectionPool
>>> p4 = HTTPSConnectionPool('ajax.googleapis.com')
>>> p4.scheme, p4.host, p4.port
('https', 'ajax.googleapis.com', None)
```

Comment (justiniso) [reply ddtrace/contrib/urllib3/patch.py]: One of the common ways of making requests via urllib3 is by using `urllib3.PoolManager().request(...)`, which abstracts the underlying connection pools from the user. As part of this abstraction, ports are assigned to the selected ConnectionPool class as a feature of the library:

https://github.com/urllib3/urllib3/blob/master/src/urllib3/poolmanager.py#L217-L221

This inferred port is finally passed to the constructor here: https://github.com/urllib3/urllib3/blob/master/src/urllib3/poolmanager.py#L268

This happens whether a port is explicitly declared in the request URL or not. 

```
>>> from ddtrace import patch
>>> patch(urllib3=True)
>>> import urllib3
>>> http = urllib3.PoolManager()
>>> r = http.request('GET', 'http://httpbin.org/robots.txt')
# In the patched function:
(Pdb) obj.port
80
```

We can monkey-patch at a higher level, detect if the port is set on the url, and set a flag on the connection pool class. But let's see if customers actually request that feature before going that route.

File: ddtrace/contrib/urllib3/patch.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
+import urllib3
+from six.moves.urllib.parse import urlparse, urlunparse
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py]: We vendor `six` and expose `parse` through `ddtrace.compat`.

File: ddtrace/contrib/urllib3/patch.py (hunk 7 above)
Comment (majorgreys) [reply ddtrace/contrib/urllib3/patch.py]: Thanks for the explanation! 👍

File: ddtrace/contrib/urllib3/patch.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    """
+    Wrapper function for the lower-level urlopen in urllib3
+
+    :param func: The original target function "urlopen"
+    :param obj: The patched instance of ``HTTPConnectionPool``
+    :param args: Positional arguments from the target function
+    :param kwargs: Keyword arguments from the target function
+    :return: The ``HTTPResponse`` from the target function
+    """
+    import pdb; pdb.set_trace()
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py]:
```suggestion
```

File: ddtrace/utils/http.py [hunk 10]
```diff
@@ -1,3 +1,6 @@
+from six.moves.urllib.parse import urlparse, urlunparse
```
Comment (majorgreys) [inline ddtrace/utils/http.py]: Also use `ddtrace.compat` here

File: ddtrace/contrib/urllib3/patch.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+            propagator = HTTPPropagator()
+            propagator.inject(span.context, request_headers)
+
+        store_request_headers(request_headers, span, config.urllib3)
+        span.set_tag(http.METHOD, request_method)
+        span.set_tag(http.URL, sanitized_url)
+        if config.urllib3["trace_query_string"]:
+            span.set_tag(http.QUERY_STRING, parsed_uri.query)
+        if config.urllib3["analytics_enabled"]:
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.urllib3["analytics_sample_rate"])
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py]: We can use the `IntegrationConfig.get_analytics_sample_rate()` function here:

```suggestion
span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.urllib3.get_analytics_sample_rate())
```

**Comment by rouge8** [comment]
Is there any chance of reviving this PR?

File: tox.ini [hunk 12]
```diff
@@ -109,6 +109,7 @@ envlist =
     tornado_contrib-py{27,35,36,37,38,39}-tornado{44,45}
     tornado_contrib-py{37,38,39}-tornado{50,51,60,}
     tornado_contrib-py27-tornado{44,45}-futures{30,31,32,}
+    urllib3_contrib-{py27,py35,py36,py37}-urllib3{122,123,124,125}
```
Comment (brettlangdon) [inline tox.ini:112]: we should add to `riotfile.py` instead, we will eventually move everything off of `tox`.

File: .circleci/config.yml [hunk 13]
```diff
@@ -920,6 +920,16 @@ jobs:
...
           pattern: '^rediscluster_contrib-'
 
+  urllib3:
+    <<: *contrib_job
+    docker:
+      - image: *ddtrace_dev_image
+      - *httpbin_local
+    resource_class: *resource_class
+    steps:
+      - run_tox_scenario:
```
Comment (brettlangdon) [inline .circleci/config.yml:930]: when we move this test to riot, we'll need to update this to:

```
- run_test:
  pattern: 'urllib3' 
```

File: .circleci/config.yml [hunk 14]
```diff
@@ -1199,6 +1209,7 @@ workflows:
       - test_logging: *requires_pre_test
       - tornado: *requires_pre_test
       - tracer: *requires_pre_test
+      - urllib3: *requires_pre_test
```
Comment (brettlangdon) [inline .circleci/config.yml:1212]: We also have to add `urllib3` to the `requires_tests` block above.

**Review by brettlangdon — COMMENTED** [review]
I only did a quick glance at the PR

File: ddtrace/utils/http.py [hunk 15]
```diff
@@ -7,3 +10,25 @@ def normalize_header_name(header_name):
     :rtype: str
     """
     return header_name.strip().lower() if header_name is not None else None
+
+
+def sanitize_url_for_tag(url):
```
Comment (Kyle-Verhoog) [inline ddtrace/utils/http.py:15]: I think we should pull this change out into a separate PR. It'd also be great to use it by default in `trace_utils.set_http_meta`.

File: ddtrace/contrib/urllib3/__init__.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
+"""
+The ``urllib3`` integration instruments tracing on http calls with optional
+support for distributed tracing across services the client communicates with.
+
+The ``patch`` function must be called before importing and using the library.
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/__init__.py:5]: We have a bit of a convention for the documentation that can be copied from the redis/django/grpc integrations

File: ddtrace/contrib/urllib3/patch.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        if config.urllib3.split_by_domain:
+            span.service = hostname
+
+        # If distributed tracing is enabled, propagate the tracing headers to downstream services
+        if config.urllib3.distributed_tracing:
+            if request_headers is None:
+                request_headers = {}
+                kwargs["headers"] = request_headers
+            propagator = HTTPPropagator()
+            propagator.inject(span.context, request_headers)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/patch.py:129]:
```suggestion
            HTTPPropagator.inject(span.context, request_headers)
```

File: ddtrace/contrib/urllib3/patch.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+    ) as span:
+        if config.urllib3.split_by_domain:
+            span.service = hostname
+
+        # If distributed tracing is enabled, propagate the tracing headers to downstream services
+        if config.urllib3.distributed_tracing:
+            if request_headers is None:
+                request_headers = {}
+                kwargs["headers"] = request_headers
+            propagator = HTTPPropagator()
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/patch.py:128]:
```suggestion
```

File: tests/contrib/urllib3/test_urllib3_patch.py [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+import pytest
+
+from ddtrace.contrib.urllib3 import patch
+from tests.contrib.patch import PatchTestCase
+
+
+pytestmark = pytest.mark.skipif("TEST_DATADOG_URLLIB3_MIGRATION" in os.environ, reason="test only without migration")
+
+
+class TestDjangoPatch(PatchTestCase.Base):
```
Comment (Kyle-Verhoog) [inline tests/contrib/urllib3/test_urllib3_patch.py:12]:
```suggestion
class TestUrllib3Patch(PatchTestCase.Base):
```

File: tests/contrib/urllib3/test_urllib3_patch.py [hunk 20]
```diff
@@ -0,0 +1,25 @@
+import os
+
+import pytest
+
+from ddtrace.contrib.urllib3 import patch
+from tests.contrib.patch import PatchTestCase
+
+
+pytestmark = pytest.mark.skipif("TEST_DATADOG_URLLIB3_MIGRATION" in os.environ, reason="test only without migration")
```
Comment (Kyle-Verhoog) [inline tests/contrib/urllib3/test_urllib3_patch.py:9]:
```suggestion
```
This line isn't needed - it was Django specific

File: tests/contrib/urllib3/test_urllib3.py [hunk 21]
```diff
@@ -0,0 +1,385 @@
...
+
+
+class TestUrllib3(BaseUrllib3TestCase):
+    def test_HTTPConnectionPool_traced(self):
+        """Tests that requests made from the HTTPConnectionPool are traced"""
+        pool = urllib3.connectionpool.HTTPConnectionPool(SOCKET)
+        # Test a relative URL
+        r = pool.request("GET", "/status/200")
+        assert r.status == 200
+        spans = self.tracer.writer.pop()
```
Comment (Kyle-Verhoog) [inline tests/contrib/urllib3/test_urllib3.py:44]:
```suggestion
        spans = self.pop_spans()
```

(also for all the other usages)

File: ddtrace/contrib/urllib3/patch.py [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+
+def unpatch():
+    """Disable trace for all urllib3 requests"""
+    if getattr(urllib3, "__datadog_patch", False):
+        setattr(urllib3, "__datadog_patch", False)
+
+        _u(urllib3.connectionpool.HTTPConnectionPool, "urlopen")
+
+
+def _infer_argument_value(args, kwargs, pos, kw, default=None):
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/patch.py:53]: This would be a good helper to have in trace_utils to be used across integrations. Might be worth doing in a separate PR but I'm ok to leave it here too. It should probably have a few unit tests as well.

File: ddtrace/contrib/urllib3/patch.py (hunk 17 above)
Comment (P403n1x87) [reply ddtrace/contrib/urllib3/patch.py:129]: 😊

File: ddtrace/utils/http.py (hunk 15 above)
Comment (P403n1x87) [reply ddtrace/utils/http.py:15]: Makes sense. I'll prepare the PR with this change

File: ddtrace/contrib/urllib3/__init__.py [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+.. py:data:: ddtrace.config.urllib3['analytics_enabled']
+
+   Whether or not to include the analytics sample rate as a tag.
+
+   Default: ``False``
+
+
+.. py:data:: ddtrace.config.urllib3['split_by_domain']
+
+   Whether or not to split by domain instead of service name.
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/__init__.py:61]:
```suggestion
   Whether or not to use the domain name of requests as the service name.
```

File: riotfile.py [hunk 24]
```diff
@@ -486,5 +486,11 @@ def select_pys(min_version=min(SUPPORTED_PYTHON_VERSIONS), max_version=max(SUPPO
                 ),
             ],
         ),
+        Venv(
+            name="urllib3",
+            pys=SUPPORTED_PYTHON_VERSIONS,
+            pkgs={"urllib3": [">=1.22,<1.27"]},
```
Comment (Kyle-Verhoog) [inline riotfile.py:492]: We should also test with the latest version of urllib3
```suggestion
            pkgs={"urllib3": [">=1.22,<1.27", latest]},
```

**Review by Kyle-Verhoog — COMMENTED** [review]
LGTM once the open comments are resolved! 😄

File: ddtrace/contrib/urllib3/patch.py [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+    Wrapper function for the lower-level urlopen in urllib3
+
+    :param func: The original target function "urlopen"
+    :param instance: The patched instance of ``HTTPConnectionPool``
+    :param args: Positional arguments from the target function
+    :param kwargs: Keyword arguments from the target function
+    :return: The ``HTTPResponse`` from the target function
+    """
+    request_method = _infer_argument_value(args, kwargs, 0, "method")
+    request_url = _infer_argument_value(args, kwargs, 1, "url")
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py:89]: These arguments are not also kwargs: 

https://github.com/urllib3/urllib3/blob/d0b20763f55536aec43caae9d180aa16c7b77d09/src/urllib3/connectionpool.py#L473-L474

```suggestion
    request_method = args[0]
    request_url = args[1]
```

File: tests/contrib/urllib3/test_urllib3.py [hunk 26]
```diff
@@ -0,0 +1,386 @@
...
+            assert "urllib3.exceptions.MaxRetryError" in s.get_tag(errors.TYPE)
+
+    def test_default_service_name(self):
+        """Test the default service name is set"""
+        out = self.http.request("GET", URL_200)
+        assert out.status == 200
+        spans = self.pop_spans()
+        assert len(spans) == 1
+        s = spans[0]
+        assert s.service == config.urllib3["_default_service"]
```
Comment (Kyle-Verhoog) [inline tests/contrib/urllib3/test_urllib3.py:208]:
```suggestion
        assert s.service == "urllib3"
```

Having the literal values helps make sure that we don't accidentally change values that users may depend on

File: ddtrace/contrib/urllib3/patch.py [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.urllib3.get_analytics_sample_rate())
+        if isinstance(request_retries, urllib3.util.retry.Retry):
+            span.set_tag(http.RETRIES_REMAIN, str(request_retries.total))
+
+        # Call the target function
+        response = None
+        try:
+            response = func(*args, **kwargs)
+        finally:
+            span.error = response is None or int(response.status >= 500)
```
Comment (Kyle-Verhoog) [inline ddtrace/contrib/urllib3/patch.py:120]:
```suggestion
```
This is handled by `set_http_meta`.

**Review by Kyle-Verhoog — COMMENTED** [review]
Just a couple tiny things but otherwise LGTM!

File: ddtrace/contrib/urllib3/patch.py (hunk 27 above)
Comment (P403n1x87) [reply ddtrace/contrib/urllib3/patch.py:120]: Ah right! 👍

**Review by Kyle-Verhoog — DISMISSED** [review]
👏

File: ddtrace/contrib/urllib3/patch.py [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+    try:
+        request_retries = get_argument_value(args, kwargs, 4, "retries")
+    except ArgumentError:
+        request_retries = None
+
+    # HTTPConnectionPool allows relative path requests; convert the request_url to an absolute url
+    if request_url.startswith("/"):
+        request_url = parse.urlunparse(
+            (
+                instance.scheme,
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py:79]: We should be defensive here and catch any `TypeError` which had been an issue with similar code in the django integration.

https://github.com/python/cpython/blob/a2f0654b0a5b4c4f726155620002cc1f5f2d206a/Lib/urllib/parse.py#L122

File: ddtrace/contrib/urllib3/patch.py [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+        if config.urllib3.distributed_tracing:
+            if request_headers is None:
+                request_headers = {}
+                kwargs["headers"] = request_headers
+            HTTPPropagator.inject(span.context, request_headers)
+
+        if config.urllib3.analytics_enabled:
+            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, config.urllib3.get_analytics_sample_rate())
+        if isinstance(request_retries, urllib3.util.retry.Retry):
+            span.set_tag(http.RETRIES_REMAIN, str(request_retries.total))
```
Comment (majorgreys) [inline ddtrace/contrib/urllib3/patch.py:113]: Should this be moved to `set_http_meta`?

**Review by majorgreys — APPROVED** [review]

File: ddtrace/contrib/urllib3/patch.py (hunk 28 above)
Comment (P403n1x87) [reply ddtrace/contrib/urllib3/patch.py:79]: Looking at https://github.com/urllib3/urllib3/blob/44ee4ad281444ec0b0a0aeeb2c8bf423c15cbc20/src/urllib3/connectionpool.py#L1004, it seems that all the URL pieces are either `None` or instances of `str`, so we should be good here.

File: ddtrace/contrib/urllib3/patch.py (hunk 29 above)
Comment (P403n1x87) [reply ddtrace/contrib/urllib3/patch.py:113]: Makes sense, but I don't think we want to treat it as a metric?

File: riotfile.py [hunk 30]
```diff
@@ -634,5 +634,11 @@ def select_pys(min_version=min(SUPPORTED_PYTHON_VERSIONS), max_version=max(SUPPO
                 ),
             ],
         ),
+        Venv(
+            name="urllib3",
+            pys=SUPPORTED_PYTHON_VERSIONS,
+            pkgs={"urllib3": [">=1.22,<1.27", latest]},
```
Comment (majorgreys) [inline riotfile.py:640]: Could we add the minimum version at least here to the matrix?

File: ddtrace/contrib/urllib3/patch.py [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+    Wrapper function for the lower-level urlopen in urllib3
+
+    :param func: The original target function "urlopen"
+    :param instance: The patched instance of ``HTTPConnectionPool``
+    :param args: Positional arguments from the target function
+    :param kwargs: Keyword arguments from the target function
+    :return: The ``HTTPResponse`` from the target function
+    """
+    request_method = get_argument_value(args, kwargs, 0, "method")
+    request_url = get_argument_value(args, kwargs, 1, "url")
```
[the remark this answers is not in the document]
Comment (P403n1x87) [reply ddtrace/contrib/urllib3/patch.py]: Whilst it's true that they are positional, it seems there's nothing there preventing a user to pass them as keyword arguments. Changing this makes an existing test fail, which to me seems valid:

~~~ python
(("POST",), {"url": URL_200, "headers": {"accept": "*"}}),
~~~

File: ddtrace/contrib/trace_utils.py [hunk 32]
```diff
@@ -189,6 +190,9 @@ def set_http_meta(
     if response_headers is not None:
         store_response_headers(dict(response_headers), span, integration_config)
 
+    if retries_remain is not None:
+        span._set_str_tag(http.RETRIES_REMAIN, str(retries_remain))
```
Comment (majorgreys) [inline ddtrace/contrib/trace_utils.py:194]: `str` is redundant since `_set_str_tag` will call `stringify/six.text_type`

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Yun-Kim — APPROVED** [review]
